expression/builtin_cast.go

```diff
@@ -0,0 +1,461 @@
+	return res, false, errors.Trace(err)
+}
+
+type builtinCastRealAsDurationSig struct {
+	baseDurationBuiltinFunc
+	fsp int
+}
+
+func (b *builtinCastRealAsDurationSig) evalDuration(row []types.Datum) (res types.Duration, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalReal(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return types.Duration{}, isNull, errors.Trace(err)
+	}
+	res, err = types.ParseDuration(strconv.FormatFloat(val, 'f', -1, 64), b.fsp)
+	return res, false, errors.Trace(err)
+}
+
+type builtinCastDecimalAsIntSig struct {
+	baseIntBuiltinFunc
+}
+
+func (b *builtinCastDecimalAsIntSig) evalInt(row []types.Datum) (res int64, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalDecimal(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	res, err = val.ToInt()
+	return res, false, errors.Trace(err)
+}
+
+type builtinCastDecimalAsStringSig struct {
+	baseStringBuiltinFunc
+}
+
+func (b *builtinCastDecimalAsStringSig) evalString(row []types.Datum) (res string, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalDecimal(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
+	}
+	return string(val.ToString()), false, nil
+}
+
+type builtinCastDecimalAsRealSig struct {
+	baseRealBuiltinFunc
+}
+
+func (b *builtinCastDecimalAsRealSig) evalReal(row []types.Datum) (res float64, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalDecimal(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	res, err = val.ToFloat64()
+	return res, false, errors.Trace(err)
+}
+
+type builtinCastDecimalAsTimeSig struct {
+	baseTimeBuiltinFunc
+	tp byte
+}
+
+func (b *builtinCastDecimalAsTimeSig) evalTime(row []types.Datum) (res types.Time, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalDecimal(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return types.Time{}, isNull, errors.Trace(err)
+	}
+	_, fsp := val.PrecisionAndFrac()
+	res, err = types.ParseTime(string(val.ToString()), b.tp, fsp)
+	return res, false, errors.Trace(err)
+}
+
+type builtinCastDecimalAsDurationSig struct {
+	baseDurationBuiltinFunc
+}
+
+func (b *builtinCastDecimalAsDurationSig) evalDuration(row []types.Datum) (res types.Duration, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalDecimal(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return types.Duration{}, false, errors.Trace(err)
+	}
+	_, fsp := val.PrecisionAndFrac()
+	res, err = types.ParseDuration(string(val.ToString()), fsp)
+	return res, false, errors.Trace(err)
+}
+
+type builtinCastStringAsIntSig struct {
+	baseIntBuiltinFunc
+}
+
+func (b *builtinCastStringAsIntSig) evalInt(row []types.Datum) (res int64, isNull bool, err error) {
+	if types.IsHybridType(b.args[0].GetType().Tp) {
```

Comment:
yeah, it's possible
For a sql like 
``` sql
select enumCol + 1;
```
The check of hybridType may be advanced to plan-phrase.
EnumCol is a hybrid type, but we don't need to add a `castStringAsRealSig` above it in plan-phrase,
we can call `enumCol.evalInt()` directly in runtime.

I'll think about, it may be better to implement it in another PR. @shenli